File: src/fillerkiller/detector/repetition.py
"""Stutter-repeat detection: "I I", "the the", and prefix repeats like "we we're".

These survive in Granola transcripts (verified), so they work on both the
post-hoc and real-time paths.
"""

from typing import TYPE_CHECKING

from fillerkiller.detector import wordlist

if TYPE_CHECKING:
    from fillerkiller.detector.engine import FillerHit

_BOUNDARY_CHARS = frozenset(".!?,;:…—–-")


def find_repetitions(
    toks: list[tuple[str, int, int]],
    text: str,
    utterance_idx: int,
) -> list["FillerHit"]:
    from fillerkiller.detector.engine import FillerHit

    hits: list[FillerHit] = []
    i = 0
    while i < len(toks) - 1:
        cur, cur_start, cur_end = toks[i]
        nxt, nxt_start, nxt_end = toks[i + 1]
        # Punctuation between the tokens is a clause or sentence boundary
        # ("tried it, it worked", "That's it. It works") or a hyphenated
        # double ("fifty-fifty") — grammar, not a stutter.
        if any(ch in _BOUNDARY_CHARS for ch in text[cur_end:nxt_start]):
            i += 1
            continue
        exact = cur == nxt and cur not in wordlist.REPETITION_ALLOW
        # Prefix repeat: "we we're", "did didn't". Require a real prefix of a
        # contraction-like continuation, not just any shared letters — and
        # respect the allowlist ("that that's" is ordinary grammar).
        prefix = (
            cur != nxt
            and cur not in wordlist.REPETITION_ALLOW
            and len(cur) >= 2
            and nxt.startswith(cur)
            and nxt[len(cur) :].startswith("'")
        )
        if exact or prefix:
            # Both copies capitalized mid-text is a proper noun ("James
            # James"), not a stutter — ASR noise clusters around names.
            # "I" is the one pronoun that's always capitalized; exempt it.
            if cur != "i" and text[cur_start].isupper() and text[nxt_start].isupper():
                i += 1
                continue
            hits.append(
                FillerHit(
                    term=text[cur_start:nxt_end].lower(),
                    category="repetition",
                    start=cur_start,
                    end=nxt_end,
                    utterance_idx=utterance_idx,
                )
            )
            i += 2  # don't re-count the second token in an overlapping pair
            continue
        i += 1
    return hits

File: macapp/FillerKillerKit/Sources/DetectorKit/Repetition.swift
// Port of src/fillerkiller/detector/repetition.py — the Python file is the
// oracle. Stutter repeats: "I I", "the the", and contraction-prefix repeats
// like "we we're".

// Punctuation between two identical tokens is a clause or sentence boundary
// ("tried it, it worked", "That's it. It works") or a hyphenated double
// ("fifty-fifty") — grammar, not a stutter. Mirrors _BOUNDARY_CHARS.
private let boundaryScalars: Set<Unicode.Scalar> = [
    ".", "!", "?", ",", ";", ":", "…", "—", "–", "-",
]

func findRepetitions(
    _ toks: [Token],
    scalars: [Unicode.Scalar],
    utteranceIdx: Int
) -> [FillerHit] {
    var hits: [FillerHit] = []
    var i = 0
    while i < toks.count - 1 {
        let cur = toks[i]
        let nxt = toks[i + 1]
        if scalars[cur.end ..< nxt.start].contains(where: { boundaryScalars.contains($0) }) {
            i += 1
            continue
        }
        let exact = cur.text == nxt.text && !Wordlist.repetitionAllow.contains(cur.text)
        var prefix = false
        if cur.text != nxt.text,
           !Wordlist.repetitionAllow.contains(cur.text),
           cur.text.unicodeScalars.count >= 2,
           nxt.text.hasPrefix(cur.text) {
            // Require a contraction continuation ("we we're"), not any prefix.
            let rest = nxt.text.dropFirst(cur.text.count)
            prefix = rest.hasPrefix("'")
        }
        if exact || prefix {
            // Both copies capitalized mid-text is a proper noun ("James
            // James"), not a stutter — ASR noise clusters around names.
            // "I" is the one pronoun that's always capitalized; exempt it.
            if cur.text != "i",
               scalars[cur.start].properties.isUppercase,
               scalars[nxt.start].properties.isUppercase {
                i += 1
                continue
            }
            hits.append(FillerHit(
                term: scalarSubstring(scalars, cur.start, nxt.end).lowercased(),
                category: "repetition",
                start: cur.start,
                end: nxt.end,
                utteranceIdx: utteranceIdx
            ))
            i += 2 // don't re-count the second token in an overlapping pair
            continue
        }
        i += 1
    }
    return hits
}
